### PR TITLE
fixes for dpos3

### PR DESCRIPTION
**Pull request by wanderingbort**

### CODE REVIEW

File: libraries/chain/include/eosio/chain/block_header_state.hpp [hunk 1]
```diff
@@ -22,10 +23,6 @@ struct block_header_state {
     public_key_type                   block_signing_key;
     vector<uint8_t>                   confirm_count;
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/block_header_state.hpp]: Move confirmations up to here.

File: libraries/chain/include/eosio/chain/block_header_state.hpp [hunk 2]
```diff
@@ -60,7 +57,7 @@ struct block_header_state {
 } } /// namespace eosio::chain
 
 FC_REFLECT( eosio::chain::block_header_state,
-            (id)(block_num)(header)(dpos_last_irreversible_blocknum)
+            (id)(block_num)(header)(dpos_irreversible_blocknum)
             (pending_schedule_lib_num)(pending_schedule_hash)
             (pending_schedule)(active_schedule)(blockroot_merkle)
             (producer_to_last_produced)(block_signing_key)
```
Comment (heifner) [inline libraries/chain/include/eosio/chain/block_header_state.hpp]: Missing confirm_count. Also I would put these in the same order as declared.

File: libraries/chain/block_header_state.cpp [hunk 3]
```diff
@@ -123,13 +106,28 @@ namespace eosio { namespace chain {
...
+          new_producer_to_last_produced[pro.producer_name] = result.dpos_irreversible_blocknum;
         }
       }
       result.producer_to_last_produced = move( new_producer_to_last_produced );
       result.producer_to_last_produced[prokey.producer_name] = result.block_num;
     }
 
+    /// grow the confirmed count
+    uint8_t required_confs = (result.active_schedule.producers.size() * 2 / 3) + 1;
+    if( confirm_count.size() < 1024 ) {
```
Comment (heifner) [inline libraries/chain/block_header_state.cpp]: Why 1024, comment would be good.

File: libraries/chain/block_header_state.cpp [hunk 4]
```diff
@@ -85,36 +85,19 @@ namespace eosio { namespace chain {
     result.block_num                             = block_num + 1;
     result.producer_to_last_produced             = producer_to_last_produced;
     result.producer_to_last_produced[prokey.producer_name] = result.block_num;
-//    result.dpos_last_irreversible_blocknum       = result.calc_dpos_last_irreversible();
+//    result.dpos_irreversible_blocknum       = result.calc_dpos_last_irreversible();
```
Comment (heifner) [inline libraries/chain/block_header_state.cpp]: Can we remove this commented out code.

**Comment by luomeiqin** [comment]
I have a question that why eos changes last_irreversible_block_num from bft_irreversible_blocknum to this ?
uint32_t controller::last_irreversible_block_num() const {
return std::max(my->head->bft_irreversible_blocknum, my->head->dpos_irreversible_blocknum);
 }
i was suppose to take bft_irreversible_blocknum is the LIB，after i go thought this issue # #3088

**Comment by BottosGroup** [comment]
This is also a problem for us

**Comment by raymondnuaa** [comment]
@luomeiqin 
Current EOS only updates dpos_irreversible_blocknum, the bft_irreversible_blocknum should be for realtime BFT. And please note there is another variable dpos_proposed_irreversible_blocknum, which is used to implemented the two rounds confirmation of BFT. (Related functions: set_confirmed and calc_dpos_last_irreversible()).

**Pull request merged**